File: bug72_wrong_data_overwrite/bug72_wrong_data_overwrite/main.java
public class main {
    public static void main(String[] args) {
        long[] arr = new long[10];
        long[] backup = new long[10];
        for (int i = 0; i < 10; i++) {
            arr[i] = 0x1111L * (i + 1);
            backup[i] = arr[i];
        }
        
        for (int i = 0; i < 10; i++) arr[i] ^= 0xDEADL;
        for (int i = 0; i < 10; i++) arr[i] ^= backup[i];
        
        long acc = 0;
        for (int i = 0; i < 10; i++) {
            acc ^= arr[i] * (i + 1) * 0x9E3779B97F4A7C15L;
        }
        
        long[] correct_arr = new long[10];
        for (int i = 0; i < 10; i++) correct_arr[i] = backup[i] ^ 0xDEADL;
        
        long correct = 0;
        for (int i = 0; i < 10; i++) {
            correct ^= correct_arr[i] * (i + 1) * 0x9E3779B97F4A7C15L;
        }
        
        if (acc == correct) {
            System.err.println("AssertionError: wrong overwrite same result");
            System.exit(1);
        }
        
        StringBuilder got = new StringBuilder("[");
        StringBuilder exp = new StringBuilder("[");
        for (int i=0; i<10; i++) {
            got.append(arr[i]).append(i==9?"":", ");
            exp.append(correct_arr[i]).append(i==9?"":", ");
        }
        got.append("]"); exp.append("]");
        
        System.err.printf("ValueError: second pass XORed with backup (original) instead of staying as-is: got %s expected %s\n", got.toString(), exp.toString());
        System.exit(1);
    }
}

File: bug72_wrong_data_overwrite/bug72_wrong_data_overwrite/main.py
import base64, struct

def run():
    arr = [0x1111 * (i + 1) for i in range(10)]
    backup = arr[:]
    for i in range(len(arr)):
        arr[i] = arr[i] ^ 0xDEAD
    for i in range(len(arr)):
        arr[i] = arr[i] ^ backup[i]
    acc = 0
    for i, v in enumerate(arr):
        acc ^= v * (i + 1) * 0x9E3779B97F4A7C15
        acc &= 0xFFFFFFFFFFFFFFFF
    original = [0x1111 * (i + 1) for i in range(10)]
    correct_arr = [v ^ 0xDEAD for v in original]
    correct = 0
    for i, v in enumerate(correct_arr):
        correct ^= v * (i + 1) * 0x9E3779B97F4A7C15
        correct &= 0xFFFFFFFFFFFFFFFF
    if acc == correct:
        raise AssertionError("wrong overwrite same result")
    raise ValueError(f"second pass XORed with backup (original) instead of staying as-is: got {arr} expected {correct_arr}")

run()
